File: BMWCenter/Core/OBD/ELM327Commands.swift
import Foundation

enum ELM327Commands {
    static let reset = "ATZ"
    static let echoOff = "ATE0"
    static let linefeedOff = "ATL0"
    static let spacesOff = "ATS0"
    static let headersOff = "ATH0"
    static let adaptiveTiming = "ATAT1"
    static let autoProtocol = "ATSP0"
    static let supportedPIDs00 = "0100"
    static let supportedPIDs20 = "0120"
    static let supportedPIDs40 = "0140"
    static let readDTCs = "03"
    static let readPendingDTCs = "07"
    static let readPermanentDTCs = "0A"
    static let clearDTCs = "04"

    static func mode01(_ pid: UInt8) -> String {
        String(format: "01%02X", pid)
    }

    /// ISO-TP header targeting DME (functional/physical request on many F-series).
    static let headerDME = "ATSH7E0"
    static let protocolCAN11_500 = "ATSP6"
    /// BMW F30/N13 MEVD1725 oil temperature (Mode 22 DID 0xD3B0).
    static let bmwOilTempMode22 = "22D3B0"

    /// Note: ATS0 (spaces off) omitted — VLinker MC-iOS / some clones mis-handle spaced replies when disabled early.
    static let initSequence: [(command: String, delayAfter: TimeInterval)] = [
        (reset, 1.5),
        (echoOff, 0.2),
        (linefeedOff, 0.2),
        (headersOff, 0.2),
        (adaptiveTiming, 0.2),
        (autoProtocol, 0.5),
        (supportedPIDs00, 0.4)
    ]
}

File: mobile/src/core/obd/elm327Commands.ts
export const ELM327Commands = {
  reset: "ATZ",
  echoOff: "ATE0",
  linefeedOff: "ATL0",
  spacesOff: "ATS0",
  headersOff: "ATH0",
  adaptiveTiming: "ATAT1",
  autoProtocol: "ATSP0",
  supportedPIDs00: "0100",
  supportedPIDs20: "0120",
  supportedPIDs40: "0140",
  readDTCs: "03",
  readPendingDTCs: "07",
  readPermanentDTCs: "0A",
  clearDTCs: "04",

  mode01(pid: number): string {
    return `01${pid.toString(16).toUpperCase().padStart(2, "0")}`;
  },

  /** ISO-TP header targeting DME (functional/physical request on many F-series). */
  headerDME: "ATSH7E0",
  protocolCAN11_500: "ATSP6",
  /** BMW F30/N13 MEVD1725 oil temperature (Mode 22 DID 0xD3B0). */
  bmwOilTempMode22: "22D3B0",
} as const;

export interface InitStep {
  command: string;
  delayAfterMs: number;
}

/** Note: ATS0 (spaces off) omitted — VLinker MC-iOS / some clones mis-handle spaced replies when disabled early. */
export const initSequence: InitStep[] = [
  { command: ELM327Commands.reset, delayAfterMs: 1500 },
  { command: ELM327Commands.echoOff, delayAfterMs: 200 },
  { command: ELM327Commands.linefeedOff, delayAfterMs: 200 },
  { command: ELM327Commands.headersOff, delayAfterMs: 200 },
  { command: ELM327Commands.adaptiveTiming, delayAfterMs: 200 },
  { command: ELM327Commands.autoProtocol, delayAfterMs: 500 },
  { command: ELM327Commands.supportedPIDs00, delayAfterMs: 400 },
];
